server/server.go

```diff
@@ -161,36 +169,68 @@ func (s *Server) Run() {
 		time.Sleep(sleep)
 	}
 
-	log.Infof("Starting server components")
-
-	s.stop = make(chan bool)
-
-	go s.Monitor()
-	go s.api.Available(s.stop)
-	go s.mach.PeriodicRefresh(machineStateRefreshInterval, s.stop)
-	go s.agent.Heartbeat(s.stop)
-	go s.aReconciler.Run(s.agent, s.stop)
-	go s.engine.Run(s.engineReconcileInterval, s.stop)
+	go s.Supervise()
 
-	beatchan := make(chan *unit.UnitStateHeartbeat)
-	go s.usGen.Run(beatchan, s.stop)
-	go s.usPub.Run(beatchan, s.stop)
+	log.Infof("Starting server components")
+	s.stopc = make(chan bool)
+	s.wg = sync.WaitGroup{}
+	beatc := make(chan *unit.UnitStateHeartbeat)
+
+	for _, f := range []func(){
+		func() { s.api.Available(s.stopc) },
+		func() { s.mach.PeriodicRefresh(machineStateRefreshInterval, s.stopc) },
+		func() { s.agent.Heartbeat(s.stopc) },
+		func() { s.aReconciler.Run(s.agent, s.stopc) },
+		func() { s.engine.Run(s.engineReconcileInterval, s.stopc) },
+		func() { s.usGen.Run(beatc, s.stopc) },
+		func() { s.usPub.Run(beatc, s.stopc) },
+	} {
+		s.wg.Add(1)
+		go func() {
+			f()
+			s.wg.Done()
+		}()
+	}
 }
 
-// Monitor tracks the health of the Server. If the Server is ever deemed
-// unhealthy, the Server is restarted.
-func (s *Server) Monitor() {
-	err := s.mon.Monitor(s.hrt, s.stop)
-	if err != nil {
+// Supervise monitors the life of the Server and coordinates its shutdown.
+// A shutdown occurs when the monitor returns, either because a health check
+// fails or a user triggers a shutdown. If the shutdown is due to a health
+// check failure, the Server is restarted. Supervise will block shutdown until
+// all components have finished shutting down or a timeout occurs; if this
+// happens, the Server will not automatically be restarted.
+func (s *Server) Supervise() {
+	err := s.mon.Monitor(s.hrt, s.killc)
+	switch {
+	case err == ErrShutdown:
+		log.Infof("Server monitor triggered: %v", err)
+		err = nil
+	case err != nil:
 		log.Errorf("Server monitor triggered: %v", err)
-
-		s.Stop()
+	default:
+		panic("unexpected nil err from Monitor")
+	}
+	close(s.stopc)
+	done := make(chan struct{})
+	go func() {
+		s.wg.Done()
+		close(done)
+	}()
+	select {
+	case <-done:
+	case <-time.After(shutdownTimeout):
+		log.Errorf("Timed out waiting for server to shut down")
+		err = nil
+	}
+	if err != nil {
+		log.Infof("Restarting server")
 		s.Run()
 	}
 }
 
+// Stop is used to gracefully terminate the server by triggering the Monitor to shut down
 func (s *Server) Stop() {
-	close(s.stop)
+	close(s.killc)
```

Comment:
If I call `server.Stop()` twice, this will still panic. Why not push something down the channel instead?